Extend parsing errors functionality and add gbif class to names data

damianooldoni opened the pull request:
This PR aims to extend parsing errors functionality to name related data (checklists). Such functionality is already present in current `rgbif` package for occurrence data. You can apply piping for filtering (in or out) occurrences with some issues and you can spread multiple issues (`split` option), expand issue (`expand` option) codes to more descriptive names or both at the same time (`split_expand` option). This means you can do the following:

```r
res <- occ_search(
      geometry = 'POLYGON((30.1 10.1, 10 20, 20 40, 40 40, 30.1 10.1))',
      limit = 50
)
# Parsing output by issue
res %>% occ_issues(gass84) # pipe operator included in rgbif package

### remove data rows with certain issue classes
res %>% occ_issues(-cudc, -cdreps)

### split issues into separate columns
res %>% occ_issues(mutate = "split")
res %>% occ_issues(-cudc, -mdatunl, -cdreps, mutate = "split")
res %>% occ_issues(gass84, mutate = "split")

### expand issues to more descriptive names
gg <- res %>% occ_issues(mutate = "expand")

### split and expand
res %>% occ_issues(mutate = "split_expand")

```

This PR extends such functionality to name related data (checklists):

```r
out <- name_usage(name = "Lupus")

# Parsing output by issue
out %>% name_issues(clasna)

### remove data rows with certain issue classes
out %>% name_issues(-bbmn, -clasna)

### split issues into separate columns
out %>% name_issues(mutate = "split")
out %>% name_issues(-scina, -clasna, mutate = "split")
out %>% name_issues(bbmn, mutate = "split")

### expand issues to more descriptive names
out %>% name_issues(mutate = "expand")

### split and expand
out %>% name_issues(mutate = "split_expand")
```

From the code point of view, this was not as straightforward as thought. So, I discussed with Scott Chamberlain the best way to do it, while maintaining the typical distinction between `occ_*` and `name_*` functions. See discussion at https://github.com/ropensci/rgbif/issues/326. Based on it, next changes have been applied:

1. the table storing the mapping of all possible issues, called `nameissues` and the function to get it, `name_issues()`, has been changed to `gbifissues()` and `gbif_issues()` respectively. In this way we can use `name_issues()` as the exact counterpart of `occ_issues()`. 
2. A column called `type` has been added to `gbif_issues()` in order to classify the issues: `type = occurrence` for occurrence related issues, `type = name` for name related issues. Again, discussed together with Scott in https://github.com/ropensci/rgbif/issues/326.
3. A call to `occ_issues()` or `name_issues()` will first get a look to the type of the issue: `occ_issues()` accepts only occurrence related issues and `name_issues()` accepts only name related issues. Informative errors are raised otherwise. This check runs in `occ_issues()` and `name_issues()` by calling the internal function `check_issues(type)`:  `check_issues(type = "occurrence", ...)` or `check_issues(type = "name", ...)` respectively.
4. as function `occ_issues()` accepts only data of class `gbif` (from `occ_search()`) or `gbif_data` (from `occ_data()`), Scott and I agreed to provide class `gbif` to name data as well (this also discussed at https://github.com/ropensci/rgbif/issues/326). A call to `name_usage()`, `name_lookup()`, `name_suggest()`, `name_backbone()` returns an object of class `gbif`. By assigning this class a better overview of the output is returned for `print(output)` or `output`, especially if the output is a list (`return = all`, the default value of `name_usage` for example). The args for the query are printed as well, the attributes (`attributes(output)`) has been added, same scheme used by Scott for occurrences. Only difference is the attribute `$type`: for occurrences can be `"single"` or `"many"` because the query is internally split in multiple queries. In case of `name_*()` functions only `"single"` is possible because the query is not split. We could also decide to not add this attribute for name related functions.
5. In order to realize (4) for functions `name_backbone` and `name_suggest` I changed list output to tibble dataframes. This makes the output of functions of this package consistent within the package: a list of data.frames (`return = all`, `verbose = TRUE`) of class `gbif` or a data.frame (`return = all`, `verbose = FALSE`) of class `gbif` as well.
6. I wrote a function `name_issues_lookup()` which is the exact counterpart of `occ_issues_lookup()`. Moreover, an improved informative error arises if you try to look up a occurrence related issue by `occ_issues_lookup()`. Example:

    ```r
    > name_issues_lookup(issue = 'CONTINENT_COUNTRY_MISMATCH')
     Error in name_issues_lookup(issue = "CONTINENT_COUNTRY_MISMATCH") : 
      Issue 'CONTINENT_COUNTRY_MISMATCH' is not related to names. Try occ_issues_lookup(issue = 
    'CONTINENT_COUNTRY_MISMATCH').
    ```

    Same with `occ_issues_lookup()` when name issues are given. More informative error handling added for some other cases as well. See files `.R/occ_issues_lookup.R` and `.R/name_issues_lookup.R` for more details about.
6. Unit-tests updated/added.
7. Documentation updated and improved.

Other small things solved:

1. Get a "real" tibble df as output of `occ_search(..., return = "data")`. There was a bug while defining the class of output: the trick is to add `'tbl'` in `class(out) <- c('tbl_df', 'tbl', 'data.frame', 'gbif')` otherwise you don't get the typical nice tibble layout of df.
2. Fixing https://github.com/ropensci/rgbif/issues/328: parsing of name related errors `in name_usage()` works now for multiple taxa as well (multiple taxa are returned by using `datasetKey` input parameter: `name_usage(datasetKey = ...)`, useful for us in TrIAS).
3. Fixing https://github.com/ropensci/rgbif/issues/335. Add `issue` to the input parameters of `name_lookup()` as this is allowed by GBIF API. I did it in order to extend the applicability of function `name_issues` to output of `name_lookup()` as well (aiming to replicate the functionalities for occurrence data where `occ_issues()` can be used on outputs of functions like `occ_search()`, `occ_data()` and download related data as well). However, I noticed that GBIF API doesn't return column `issues`, so we cannot apply any filtering on such data via `name_issues()`, a bug of GBIF API. I documented it in https://github.com/gbif/gbif-api/issues/37.

damianooldoni committed:
typo

damianooldoni committed:
Update list of issues  gbifissues

The issues in nameissues are merged into gbifissues.

damianooldoni committed:
Update source adding link to nameusage issues

damianooldoni committed:
seealso not more needed

damianooldoni committed:
Add type field to documentation

damianooldoni committed:
Add word occurrence

damianooldoni committed:
Add default NULL value to issue parameter

damianooldoni committed:
add parsed issues to args

damianooldoni committed:
Add issue parameter t oname_lookup

This implements the same functionality of occ_search for names instead of occurrences

damianooldoni committed:
remove `issue` from the parameter

This is not supported by API: api.gbif.org/v1/species?.
It is supported by api.gbif.org/v1/species/search? (rgbif::name_lookup() function)

damianooldoni committed:
Handle issues

From list to character string made of codes

damianooldoni committed:
typos

damianooldoni committed:
Moving parsing issues function to zzz

damianooldoni committed:
Move parsing issues to zzz

So, remove it from occ_search()

damianooldoni committed:
Ignore Rproject files

damianooldoni committed:
Improve documentation

Description of expand should be placed above the description of split_expand. Add period at the end of description's sentences.

damianooldoni committed:
Apply gbif S3 class to names

damianooldoni committed:
Set type equal to single for output of name_usage()

name_usage doesn't accept multiple values of input params.

damianooldoni committed:
Add check of start value

This check was not present but it should be here. Same as for occurrences.

damianooldoni committed:
Add explicative comment about attribute type

name_usage output is always of type single

damianooldoni committed:
Remove argscoll

argscoll is not needed: name_usage triggers a single query so args is sufficient.

damianooldoni committed:
Small changes to print_gbif function for printing names

Changes applied to part with type = "single"(names are retrieved without splitting the parameter values, so type is always single for name_* functions).

damianooldoni committed:
Improve uniformity

Put records as for print when type = "single"

damianooldoni committed:
Change last part of name_lookup to return a gbif class variable

damianooldoni committed:
Make of name_issues the same function of occ_issues

Documentation is wtill work in progress

damianooldoni committed:
Add names in printing gbif class

names is returned by name_lookup()

damianooldoni committed:
Move biggest part of issues handling code in zzz.R

As heart and ancillary functions are used by name_issues() as well, they have to be placed in zzz.R

damianooldoni committed:
name_issues()

damianooldoni committed:
Update documentation

you can query based on issues via name_lookup() at the moment (see https://github.com/gbif/gbif-api/issues/37) but you don't get any information about issues back. So, this part of documentation (copy from occ_issues) should be removed.

damianooldoni committed:
code style / typo

damianooldoni committed:
Start changing unit-tests

damianooldoni committed:
Merge branch 'master' of https://github.com/damianooldoni/rgbif into add_parsing_name_issues

damianooldoni committed:
Adapt occ_issues_lookup

damianooldoni committed:
Add name_issues_lookup

The name counterpart of occ_issues_lookup

damianooldoni committed:
Add tests for *_issues_lookup

Both occ_issues_lookup and name_issues_lookup

damianooldoni committed:
Adapt documentation occ_issues

damianooldoni committed:
devtools::document()

damianooldoni committed:
Add examples to documentation

damianooldoni committed:
Add parameter is_occ to params of help funcs handle_issues() and split_iss

damianooldoni committed:
Remove too much empty lines at the end

damianooldoni committed:
typos examples documentation occ_issues

damianooldoni committed:
Update tests name_usage about type of output (gbif instead of list))

damianooldoni committed:
Update name_issues() based on new functionality

name_issues() is the parallel of occ_issues() for names. No more list of issues: that is now called gbif_issues().

damianooldoni committed:
remove spaces

damianooldoni committed:
devtools::document()

damianooldoni committed:
Change name of txt file with table of all GBIF issues

Ii don't have it on my laptop, vrey likely a txt file on laptop of Scott... Still, the name refers obviously to such table.

damianooldoni committed:
Use of is_true() is deprecated

This was a warning while testing. Solved now.

damianooldoni committed:
name_lookup gives gbif objects back not list anymore

Same as name_usage

damianooldoni committed:
split on more lines

damianooldoni committed:
solve bug while returning meta

damianooldoni committed:
Split on more rows

damianooldoni committed:
Replave ' with " in examples name_lookup

Add coherency to use of " ' and ' ' in examples

damianooldoni committed:
Improve consistency in documentation

Consistent end sentences with full stop.
Consistent use of ' ' as Scott has used in other functions.

damianooldoni committed:
Improve documentation name_usage

Output description: ame scheme as in name_lookup.

damianooldoni committed:
Solve bug with hierarcy and names in name_lookup

heirarchy and names were not returned if return  = "hierarchy" or return = "names". But they were ok if return = "all". Problem in the if statement

damianooldoni committed:
Return output of occ_search dataframes as tibble dataframes

damianooldoni committed:
Correct typo

damianooldoni committed:
Update tests occ_search

damianooldoni committed:
Put as_data_fame back again

the problem was the absence of "tbl" in class()

damianooldoni committed:
Update documentation function name_suggest

damianooldoni committed:
Add 'tbl' to class output

Otherwise the print of df will be still old shool df

damianooldoni committed:
Update and imrpove unit-test name_suggest

damianooldoni committed:
Add gbif class and structure to output name_backbone

damianooldoni committed:
Add unit-tests to test-name_backbone

Test changes to class and structure

damianooldoni committed:
Add tests to test-name_lookup

damianooldoni committed:
Put "as_data_frame" back: "as_tibble" not needed

damianooldoni committed:
Remove wrong subsetting: tt is already a df

damianooldoni committed:
Update documentation name_backbone

damianooldoni committed:
remoe typo in name cassette name_backbone

damianooldoni committed:
Imrpove output of name_backbone and update documentation

damianooldoni committed:
Reduce length title doc name_suggest

Description should contain a long description, title should be short.

damianooldoni committed:
Merge master from rgbif solving conflicts

Conflicts in occ_issues and its documentation solved.

peterdesmet committed:
Use period for function description

peterdesmet committed:
Use code syntax for columns

peterdesmet committed:
Update title and description for gbif_issues

@sckott you are probably aware of this, but if you leave an empty line between the sentence "Table has the following fields:" and the list, that roxygen will consider the sentence = @description, and the list = @details, resulting them to be separated. See for example: https://ropensci.github.io/rgbif/reference/gbif_issues.html

To keep the description as a whole, it cannot contain empty lines.

peterdesmet reviewed (approved):
Nice work @damianooldoni 

I updated the issue body (making some markdown corrections) so it can be used for the final PR + I updated the description of `gbif_issues`.

damianooldoni commented:
Thanks to improve the issue body as well. In case the content of PR doesn't change considerably after review of @stijnvanhoey, we can reuse it for final PR, indeed.

stijnvanhoey commented on R/name_backbone.r:
```diff
@@ -76,10 +78,17 @@ name_backbone <- function(name, rank=NULL, kingdom=NULL, phylum=NULL,
+                 stringsAsFactors = FALSE))
+    out <- list(data = dat,
+                alternatives = alt)
+    class(out) <- "gbif"
```
Is there any particular reason to use a `gbif` specific class? What is the difference with just a list? Ar you adding specific attributes,methods or checks? If not and it is just a list, than use `list`.  And if you add specific methods/... is `gbif` than a proper class name?

stijnvanhoey replied:
after quickly scanning https://github.com/damianooldoni/rgbif/pull/1/files#diff-4a8742c6251b78d73613f1ce59c27322 and the issue https://github.com/ropensci/rgbif/issues/326 I see you have a special `print` method for the `gbif` class, so I understand the reasoning. :-)

sckott replied:
@stijnvanhoey right, the print method for nicer user interface

stijnvanhoey commented on R/name_backbone.r:
```diff
@@ -76,10 +78,17 @@ name_backbone <- function(name, rank=NULL, kingdom=NULL, phylum=NULL,
   } else {
-    structure(tt[!names(tt) %in% c("alternatives", "note")], note = tt$note)
+    out <- tibble::as_data_frame(tt[!names(tt) %in% c("alternatives", "note")])
+    class(out) <- c('tbl_df', 'tbl', 'data.frame', 'gbif')
```
cfr. `gbif` class comment.

stijnvanhoey commented on R/name_issues_lookup.r:
```diff
@@ -0,0 +1,57 @@
+    gbifissues[grep(issue, gbifissues$issue),]
+  } else{
+    if (!code %in% gbifissues$code) {
+      stop("Invalid code.")
```
Invalid or not existing?

sckott replied:
this could be clarified, i guess it means the code provided doesn't exist in our set of codes. the error message could be better

damianooldoni replied:
I change it to `"Issue code not exists."`

stijnvanhoey commented on R/name_issues.R:
```diff
@@ -1,74 +1,58 @@
+#' aa %>% name_issues(-clasna, -scina)
+#'
+#' ### expand issues to more descriptive names
+#' aa %>% name_issues(mutate = "expand")
```
I do understand this feature is not new and actually inspired by the existing `occ_issues` setup, but is this not misusing the `mutate` verb? I would expect `mutate` to be a function instead of an argument? Or I would intuitively expect a column with all values equal to `expand` interpreting this code without context ( where it actually is `mutate("expand")`).  Is this not making readability of the code harder? 

I would propose to change the argument name `mutate`

damianooldoni replied:
I found it also strange as I come from `tidyverse` world. But, documentation was sufficiently good to unnderstand the difference. I didn't to change this as well as not part of my initial PR. @sckott: something to think about, but maybe not for this PR?

sckott replied:
the very immedate problem comes if you introduce a `rgbif::mutate` function that it clashes with `dplyr::mutate`, yes? is there a param name that is more clear than mutate?

stijnvanhoey commented on R/occ_issues_lookup.r:
```diff
@@ -1,17 +1,59 @@
 
 occ_issues_lookup <- function(issue=NULL, code=NULL){
-  if(is.null(code))
+
```
As you merged the gbifissues table as a single table, the functionality here `occ_issues_lookup` and `name_issues_lookup` are very similar. I would [DRY](https://en.wikipedia.org/wiki/Don%27t_repeat_yourself) this and abstract away the issue lookup as a single function (`gbif_issues_lookup`)

damianooldoni replied:
Thanks @stijnvanhoey : nice idea.

I did it just because the function `occ_issues_lookup` already exists in `rgbif` in file [`occ_issues_lookup.r`](https://github.com/ropensci/rgbif/blob/master/R/occ_issues_lookup.r) and it matches nice with `occ_issues()` in [`occ_issues.r`](https://github.com/ropensci/rgbif/blob/master/R/occ_issues.r). As @sckott and I decided to have `name_issues` as counterpart of `occ_issues`, I thought to maintain the difference between _name-_ and _occ-_ functions also the `*_lookup` one. I tried to limit as much as possible the repetitions by using the common function `check_issues` in `zzz.r`. 

However, `gbif_issues_lookup` would form a nice pair with `gbif_issues`: the first returning the details of the issue (type: `name` or `occurrence` included) the last one returning all table.

So, a question for @sckott : how would you like to have this implemented? I vote for @stijnvanhoey's idea.

sckott replied:
Merging seems like a good idea

damianooldoni replied:
Finally I will work on this now: this is the only issue which avoids to push the PR a step further. Other comments are general issues/discussions.. Sorry for months delay.

damianooldoni replied:
See a35a1bfa2862861099f3b47e0b28a95f81fa4121

stijnvanhoey commented on R/print_gbif.R:
```diff
@@ -1,14 +1,33 @@
-    cat(rgbif_wrap(sprintf("No. unique hierarchies [%s]", length(x$hierarchy))), "\n")
-    cat(rgbif_wrap(sprintf("No. media records [%s]", length(x$media))), "\n")
-    cat(rgbif_wrap(sprintf("No. facets [%s]", length(x$facets))), "\n")
+    if (is.null(attr(x, "type"))) FALSE else attr(x, "type") == "single"
```
The set of `if` statements inside the conditional section of the main `if` is rather difficult to interpret. 

Although I understand the case of this custom `print`, it feels a bit hacky due to the wide variety of forms the `gbif` object actually can have (`data.frame`, list of `data.frame`s each with varying attributes)... As the main use is to support `print` (polymorphisme), why not create different clearly defined classes `gbif_xxx` each with a easy to write custom print functionality? Adding support for new or adapted `gbif`-like objects would result in far less spaghetti-like code, I think. Still, a matter of choice and maybe not worth the effort to split everything.

damianooldoni replied:
I had the same feeling; @stijnvanhoey : I started doing it and I realized that it was not really an easy task to make it work for names too. However, the difference in list of dataframes and a single dataframe was alread present. @sckott : what do you think about?

sckott replied:
i can see the case for separate S3 classes, seems better to open as a separate issue in the main repo

stijnvanhoey reviewed (commented):
First of all @damianooldoni nice extension of the package in line with the existing functionalities. Good work.

I have some general comments on style and setup of the code, some of which are probably comments on existing code of `rgbif` instead of your code specific. Feel free to ignore these.

peterdesmet commented:
@stijnvanhoey, nice in-depth review, but touches on design decisions made for rgbif in general. I think they are relevant, but

1. I would let Scott weigh in. @damianooldoni, this is an “internal” PR, but might be worth to include him in?
2. I don’t think they block the PR for this particular feature?

damianooldoni commented:
Thanks @peterdesmet and @stijnvanhoey  for reviewing! I agree with all points actually. As @peterdesmet said, ideas proposed by @stijnvanhoey are quite general and the opinion of @sckott is needed.

sckott commented:
replied to the comments where i was mentioned

sckott committed:
#348 fixes for occ_download_queue to fail better when GBIF credentials not given

sckott committed:
fix #350 replace all tibble as_Data_frame and data_Frame with as_tibble

sckott committed:
fix #347 occ_download fail better when use does not give string inputs to ...

sckott committed:
bump version

sckott committed:
fix #349 - make gbifID and key fields in occ_search/occ_data/occ_get character class

sckott committed:
#352 add format parameter to occ_download fxn
allow specifying download format (dwca, csv or species list) for download requests
bump version

sckott committed:
fix download_parsing tests for new format param #352

sckott committed:
#352 fill out download_parsing test for species_list and simple_csv format options

sckott committed:
fix to gbif_citation - convert keys to numeric internally
due to change in occ_search/occ_data to return character class keys most likely

sckott committed:
fix #351 drop NA as well as NULL with internal fxn rgbif_compact
bump version

sckott committed:
fix #355 use preserve exact bytes for many fixtures that caused debian clang errors on cran checks

sckott committed:
fix elevation expect_error expectation

sckott committed:
bump some min. versions and update readme.md

sckott committed:
update revdep

sckott committed:
update news and cran comments

sckott committed:
update docs

sckott committed:
fix length 1 only conditional

sckott committed:
clean up contributing.md

sckott committed:
fix links in readme

sckott committed:
update docs

sckott committed:
fix url in readme for real

sckott committed:
update codemeta.json file

sckott committed:
bump to dev version

damianooldoni committed:
Merge branch 'master' of https://github.com/damianooldoni/rgbif into add_parsing_name_issues

damianooldoni committed:
Merge remote-tracking branch 'upstream/master' into add_parsing_name_issues

damianooldoni committed:
Merge name_issues_lookup and occ_issues_lookup

See: https://github.com/damianooldoni/rgbif/pull/1#discussion_r262737013

damianooldoni committed:
Update documentation

Replace doc of the two *_issues_lookup functions with gbif_issues_lookup

damianooldoni committed:
Correct typos

damianooldoni committed:
Correct typo in cassette name

damianooldoni committed:
Merge tests name/occ_issues_lookup into gbif_issues_lookup

damianooldoni committed:
Improve error message gbif_issues_lookup

damianooldoni committed:
Correct small bug/typo

damianooldoni committed:
Add documentation about added param "issue"

Related to https://github.com/ropensci/rgbif/issues/335

damianooldoni committed:
Remove errors in test-occ_search

These errors occur in official rgbif master too.
And set name variables back as in rgbif@master otherwise we get problems with vcr cassette.

damianooldoni committed:
Add small changes in documentation

damianooldoni commented:
Required changes done. 
One test at `test-elevation` doesn't pass. However, it is not part of this PR. Same error arises while installing `rgbif` from `ropensci/rgbif@master`. I will write more details about in an issue.
